### PR TITLE
PresentationActionType and SelectionPlan many-to-many relation support

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3084239&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Models/Foundation/Summit/Events/Presentations/Actions/AllowedPresentationActionType.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+use models\utils\One2ManyPropertyTrait;
+use models\utils\SilverstripeBaseModel;
+use Doctrine\ORM\Mapping AS ORM;
+/**
+ * @ORM\Entity
+ * @ORM\Table(name="PresentationActionType_SelectionPlan")
+ * Class AllowedPresentationActionType
+ * @package models\summit
+ */
+class AllowedPresentationActionType extends SilverstripeBaseModel implements IOrderable
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/Presentations/Actions/AllowedPresentationActionType.php:26]: inherit from BaseEntity ( SSBase model provide classType, and timestamp that does not needed on a join table)

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php [hunk 2]
```diff
@@ -46,6 +47,22 @@ class PresentationActionType extends SilverstripeBaseModel
      */
     private $order;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:48]: add deprecated anotaton

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php [hunk 3]
```diff
@@ -46,6 +47,22 @@ class PresentationActionType extends SilverstripeBaseModel
...
+     */
+    private $selection_plans;
+
+    /**
+     * PresentationActionType constructor.
+     */
+    public function __construct()
+    {
+        parent::__construct();
+        $this->order = 1;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:62]: remove this

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php [hunk 4]
```diff
@@ -46,6 +47,22 @@ class PresentationActionType extends SilverstripeBaseModel
      */
     private $order;
 
+    /**
+     * @ORM\OneToMany(targetEntity="models\summit\AllowedPresentationActionType", mappedBy="type", cascade={"persist"}, orphanRemoval=true, fetch="EXTRA_LAZY")
+     * @var AllowedPresentationActionType[]
+     */
+    private $selection_plans;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:54]: collection should be called something more meaningful like assigned_selection_plans

File: app/Models/Foundation/Summit/SelectionPlan.php [hunk 5]
```diff
@@ -974,4 +985,72 @@ public function setSubmissionLockDownPresentationStatusDate(?DateTime $submissio
...
+        $criteria->orderBy(['order' => 'DESC']);
+        $res = $this->presentation_action_types->matching($criteria)->first();
+        return $res === false ? 0 : $res->getOrder();
+    }
+
+    /**
+     * @param PresentationActionType $presentation_action_type
+     */
+    public function addPresentationActionType(PresentationActionType $presentation_action_type) {
+        $order = $this->getPresentationActionTypesMaxOrder();
```
Comment (smarcet) [inline app/Models/Foundation/Summit/SelectionPlan.php:1012]: u need to check here if the PresentationActionType is already assigned or not using a query DQL

File: database/migrations/model/Version20221101142824.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        $builder = new Builder($schema);
+
+        if (!$schema->hasTable("PresentationActionType_SelectionPlan")) {
+
+            $builder->create('PresentationActionType_SelectionPlan', function (Table $table) {
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+                $table->timestamp('Created');
+                $table->timestamp('LastEdited');
```
Comment (smarcet) [inline database/migrations/model/Version20221101142824.php:37]: u dont need this

File: database/migrations/model/Version20221101142824.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    public function up(Schema $schema): void
+    {
+        $builder = new Builder($schema);
+
+        if (!$schema->hasTable("PresentationActionType_SelectionPlan")) {
+
+            $builder->create('PresentationActionType_SelectionPlan', function (Table $table) {
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+                $table->timestamp('Created');
```
Comment (smarcet) [inline database/migrations/model/Version20221101142824.php:36]: u dont need this

File: app/Models/Foundation/Summit/SelectionPlan.php [hunk 8]
```diff
@@ -974,4 +985,72 @@ public function setSubmissionLockDownPresentationStatusDate(?DateTime $submissio
...
+        $this->presentation_action_types->removeElement($presentation_action_type_assignment);
+        self::resetOrderForSelectable($this->presentation_action_types, AllowedPresentationActionType::class);
+    }
+
+    /**
+     * @param PresentationActionType $presentation_action_type
+     * @param int $order
+     * @throws ValidationException
+     */
+    public function updateOrder(PresentationActionType $presentation_action_type, int $order) {
```
Comment (smarcet) [inline app/Models/Foundation/Summit/SelectionPlan.php:1034]: take a look on the same class here
https://github.com/OpenStackweb/summit-api/blob/417557f69bc7a3751993f39a84a44a02cb534750/app/Models/Foundation/Summit/SelectionPlan.php#L654
please lets keeps the naming convention of the method

File: app/Models/Foundation/Summit/Events/Presentations/Actions/AllowedPresentationActionType.php (hunk 1 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/Presentations/Actions/AllowedPresentationActionType.php:26]: done

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:48]: done

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php (hunk 3 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:62]: done

File: app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php (hunk 4 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Events/Presentations/Actions/PresentationActionType.php:54]: done

File: database/migrations/model/Version20221101142824.php (hunk 6 above)
Comment (romanetar) [reply database/migrations/model/Version20221101142824.php:37]: removed

File: database/migrations/model/Version20221101142824.php (hunk 7 above)
Comment (romanetar) [reply database/migrations/model/Version20221101142824.php:36]: removed

File: app/Models/Foundation/Summit/SelectionPlan.php (hunk 8 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SelectionPlan.php:1047]: done

File: app/Models/Foundation/Summit/SelectionPlan.php (hunk 5 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/SelectionPlan.php:1025]: done

**Comment by romanetar** [comment]
@smarcet every review item was resolved in the last commit

**Comment by smarcet** [comment]
@romanetar 
u need to update https://github.com/OpenStackweb/summit-api/blob/feature/progress-flags-by-submission-plan/app/ModelSerializers/Summit/SelectionPlanSerializer.php

to expose the collection allowed_presentation_action_types

also we would need to refactor this https://github.com/OpenStackweb/summit-api/blob/cda3f37b66aa4a3563e2f1be6864e7b05006d797/app/Services/Model/Imp/SummitPresentationActionService.php#L34
to do it by action type id and not action id ( if the action of the provide type does not exist create it new PresentationAction )
and remove all this logic
https://github.com/OpenStackweb/summit-api/blob/cda3f37b66aa4a3563e2f1be6864e7b05006d797/app/Models/Foundation/Summit/Events/Presentations/Presentation.php#L1614
https://github.com/OpenStackweb/summit-api/blob/cda3f37b66aa4a3563e2f1be6864e7b05006d797/app/Models/Foundation/Summit/Summit.php#L5703

https://github.com/OpenStackweb/summit-api/blob/feature/progress-flags-by-submission-plan/app/Jobs/SynchPresentationActions.php
and 
https://github.com/OpenStackweb/summit-api/blob/feature/progress-flags-by-submission-plan/app/Jobs/SynchAllPresentationActions.php

also review this serializer

https://github.com/OpenStackweb/summit-api/blob/cda3f37b66aa4a3563e2f1be6864e7b05006d797/app/ModelSerializers/Summit/Presentation/PresentationSerializer.php#L202

getPresentationActions should filter the ones that belongs to the current selection plan

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments
regards

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php [hunk 9]
```diff
@@ -1097,4 +1107,151 @@ public function detachEventType($id, $selection_plan_id, $event_type_id)
...
+     */
+    public function getAllowedPresentationActionType($summit_id, $selection_plan_id, $type_id) {
+        return $this->processRequest(function() use($summit_id, $selection_plan_id, $type_id) {
+            $summit = SummitFinderStrategyFactory::build($this->summit_repository, $this->resource_server_context)->find($summit_id);
+            if (is_null($summit)) return $this->error404();
+
+            $selection_plan = $summit->getSelectionPlanById(intval($selection_plan_id));
+            if (is_null($selection_plan)) return $this->error404();
+
+            $presentation_action_type = $summit->getPresentationActionTypeById(intval($type_id));
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php:1186]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php [hunk 10]
```diff
@@ -1097,4 +1107,151 @@ public function detachEventType($id, $selection_plan_id, $event_type_id)
...
+
+            $allowed_presentation_action_type = $selection_plan->getPresentationActionType($presentation_action_type);
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php:1188]: done

File: database/migrations/model/Version20221101142824.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    public function up(Schema $schema): void
+    {
+        $builder = new Builder($schema);
+
+        if (!$schema->hasTable("PresentationActionType_SelectionPlan")) {
+
+            $builder->create('PresentationActionType_SelectionPlan', function (Table $table) {
+                $table->integer("ID", true, false);
+                $table->primary("ID");
+                $table->integer('CustomOrder')->setNotnull(true)->setDefault(0);
```
[the remark this answers is not in the document]
Comment (romanetar) [reply database/migrations/model/Version20221101142824.php:36]: done

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php [hunk 12]
```diff
@@ -34,6 +38,15 @@ protected function getBaseEntity()
        return PresentationActionType::class;
     }
 
+    /**
+     * @param QueryBuilder $query
+     * @return QueryBuilder
+     */
+    protected function applyExtraJoins(QueryBuilder $query, ?Filter $filter = null){
+        $query = $query->leftJoin("e.assigned_selection_plans", 'sp', Join::ON);
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:46]: as we discussed, it remains as it is

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php [hunk 13]
```diff
@@ -52,7 +65,7 @@ protected function getOrderMappings()
     {
         return [
             'id' => 'e.id',
-            'order' => 'e.order',
+            'order' => 'sp.order',
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:68]: as we discussed, it remains as it is

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php [hunk 14]
```diff
@@ -1097,4 +1107,151 @@ public function detachEventType($id, $selection_plan_id, $event_type_id)
...
+    {
+        return $this->processRequest(function () use ($id, $selection_plan_id, $type_id) {
+            $summit = SummitFinderStrategyFactory::build($this->summit_repository, $this->resource_server_context)->find($id);
+            if (is_null($summit)) return $this->error404();
+
+            $payload = $this->getJsonPayload(SummitSelectionPlanValidationRulesFactory::buildForAddPresentationActionType());
+
+            $this->selection_plan_service->upsertAllowedPresentationActionType(
+                $summit, intval($selection_plan_id), intval($type_id), $payload);
+            return $this->created();
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitSelectionPlansApiController.php:1217]: Done. It returns a PresentationActionType with the SelectionPlan assignment order. Keep in mind that from now on the order property in PresentationActionType is the assignment order so it's not needed to return the association entity (AllowedPresentationActionType)

File: app/Models/Foundation/Summit/SelectionPlan.php [hunk 15]
```diff
@@ -974,4 +985,72 @@ public function setSubmissionLockDownPresentationStatusDate(?DateTime $submissio
     public function clearSubmissionLockDownPresentationStatusDate():void{
         $this->submission_lock_down_presentation_status_date = null;
     }
+
+    /**
+     * @return ArrayCollection
+     */
+    public function getPresentationActionTypes(): ArrayCollection {
+        return $this->presentation_action_types->map(function ($entity) {
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Models/Foundation/Summit/SelectionPlan.php]: I return a PresentationActionType (as the method signature says) here and then the serializer resolves the assignment order  here: https://github.com/OpenStackweb/summit-api/blob/cda3f37b66aa4a3563e2f1be6864e7b05006d797/app/ModelSerializers/Summit/Presentation/PresentationActionTypeSerializer.php#L43

I think that both are valid approaches but this one is already tested :)

**Comment by romanetar** [comment]
@smarcet last commit has been rebased with main and contains all the reviews fixes. 
Please check out the last review's replies and let me know if u agree.

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php [hunk 16]
```diff
@@ -29,30 +33,41 @@ final class DoctrinePresentationActionTypeRepository
...
     }
 
+    /**
+     * @param QueryBuilder $query
+     * @param Filter|null $filter
+     * @return QueryBuilder
+     */
+    protected function applyExtraJoins(QueryBuilder $query, ?Filter $filter = null): QueryBuilder
+    {
+        return $query->leftJoin("e.assigned_selection_plans", 'sp', Join::ON);
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:48]: this is not correct
u should be doing

$query->leftJoin("e.assigned_selection_plans", 'asp')->lefJoind('asp.selection_plan", "sp");

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php [hunk 17]
```diff
@@ -29,30 +33,41 @@ final class DoctrinePresentationActionTypeRepository
...
      * @return array
      */
-    protected function getFilterMappings()
+    protected function getFilterMappings(): array
     {
         return [
             'label' => 'e.label:json_string',
-            'summit_id' => new DoctrineLeftJoinFilterMapping("e.summit", "s" ,"s.id :operator :value")
+            'summit_id' => new DoctrineLeftJoinFilterMapping("e.summit", "s" ,"s.id :operator :value"),
+            'selection_plan_id' => new DoctrineLeftJoinFilterMapping("e.assigned_selection_plans", "sp" ,"sp.selection_plan_id :operator :value")
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:59]: u dont need this join
u are already doing the join at applyExtraJoins
just do
 'selection_plan_id' => 'sp.id'

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php [hunk 18]
```diff
@@ -29,30 +33,41 @@ final class DoctrinePresentationActionTypeRepository
...
     /**
      * @return array
      */
-    protected function getOrderMappings()
+    protected function getOrderMappings(): array
     {
         return [
             'id' => 'e.id',
-            'order' => 'e.order',
+            'order' => 'sp.order',
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:70]: this should be asp.order check the fix for applyExtraJoins

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments on code
and the main comments

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php (hunk 18 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:70]: done

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php (hunk 17 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:59]: done

File: app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php (hunk 16 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrinePresentationActionTypeRepository.php:48]: done

**Comment by romanetar** [comment]
@smarcet last commit has been rebased with main and contains all the reviews fixes including the main one.

**Pull request merged**